2017/proposals/bill-engels-pymc3-proposal.md

```diff
@@ -0,0 +1,292 @@
+# Explore Alternative Computation Engines for PyMC3
+
+## Abstract
+
+Theano provides symbolic differentiation of mathematical expressions, enabling
+the application of powerful, gradient-based inference algorithms to arbitrary
+probabilistic models in PyMC3.  Vectorization, parallelism, GPU computation,
+and dynamic C code generation are added benefits.  Underlying the deep learning
+boom, other computational frameworks have also become popular, such as
+TensorFlow, MxNet, CNTK, Chainer, PyTorch, and others.  Perhaps the most
+relevant of these to PyMC3 are TensorFlow and PyTorch.  Like Theano, these
+frameworks include extensive libraries of low-level mathematical and linear
+algebra operations.  
+
+The goal of this work will be to examine how alternative computation engines
+effect the performance and the ease of use of PyMC3.  I will make two forks of
+PyMC3, one with TensorFlow and the other with PyTorch in place of Theano.  I
+will analyze the performance of common models and tests in each of the two
+forks and the master.  Via blog posts, I will document performance results with
+an eye towards how the backend impacts ease-of-use from the perspective of a
+user.  This work will also result in a detailed programming guide and PyMC3
+architecture overview (see [issue #1359][issue]) that can be added to the
+documentation.  Finally, my mentors and I will design an abstraction layer
+between PyMC3 core code and Theano.  If deemed successful, I hope to merge this
+into master by the end of Phase 3.
+
```

Comment:
Another potential benefit is that by studying the theano architecture of PyMC3, we might also find optimizations there.